filestack/models/filestack_client.py

```diff
@@ -116,48 +117,75 @@ def upload(self, url=None, filepath=None, multipart=True, params=None, upload_pr
+                response = intelligent_ingestion.upload(
+                    self.apikey, filepath, self.storage, params=params, security=self.security
+                )
+
+            elif multipart:
+                response = upload_utils.multipart_upload(
+                    self.apikey, filepath, self.storage,
+                    upload_processes=upload_processes, params=params, security=self.security
+                )
+                handle = response['handle']
+                return filestack.models.Filelink(handle, apikey=self.apikey, security=self.security)
+
+            else:  # Uploading with multipart=False
                 filename = os.path.basename(filepath)
                 mimetype = mimetypes.guess_type(filepath)[0]
                 files = {'fileUpload': (filename, open(filepath, 'rb'), mimetype)}
 
-            if params:
-                params['key'] = self.apikey
-            else:
-                params = {'key': self.apikey}
+                if params:
+                    params['key'] = self.apikey
+                else:
+                    params = {'key': self.apikey}
+
+                path = '{path}/{storage}'.format(path=STORE_PATH, storage=self.storage)
 
-            path = '{path}/{storage}'.format(path=STORE_PATH, storage=self.storage)
+                if self.security:
+                    path = "{path}?policy={policy}&signature={signature}".format(
+                        path=path, policy=self.security['policy'].decode('utf-8'),
+                        signature=self.security['signature']
+                    )
+
+                response = utils.make_call(
+                    API_URL, 'post', path=path, params=params, files=files
+                )
+
+        else:  # Uploading from an external URL
+            tasks = []
+            request_url_list = []
+
+            if params:
+                store_task = utils.store_params_maker(params)
+                tasks.append(store_task)
 
             if self.security:
-                path = "{path}?policy={policy}&signature={signature}".format(
-                    path=path, policy=self.security['policy'].decode('utf-8'),
-                    signature=self.security['signature']
+                tasks.append(
+                    'security=p:{policy},s:{signature}'.format(
+                        policy=self.security['policy'].decode('utf-8'),
+                        signature=self.security['signature']
+                    )
                 )
 
-            response = utils.make_call(
-                API_URL, 'post', path=path, params=params, data=data, files=files
-            )
+            tasks = '/'.join(tasks)
+
+            if tasks:
+                request_url_list.extend((CDN_URL, self.apikey, tasks, url))
+            else:
+                request_url_list.extend((CDN_URL, self.apikey, url))
+
+            request_url = '/'.join(request_url_list)
+
+            response = requests.post(request_url)
```

Comment:
all requests should be made with `config.HEADERS`

Reply:
Done!